pkg/services/hivelocity/device/device.go

```diff
@@ -393,3 +343,134 @@ func createTags(clusterName, machineName string, isControlPlane bool) []string {
+	deviceTags := append(device.Tags, clusterAndMachineTag(s.scope.HivelocityCluster.Name, s.scope.Name())...)
+	if err := s.scope.HVClient.SetTags(ctx, device.DeviceId, deviceTags); err != nil {
+		return fmt.Errorf("failed to set tags on machine %s :%w", s.scope.Name(), err)
+	}
+	// Set update tags of device to have the latest device for the next steps in reconcile
+	device.Tags = deviceTags
+	return nil
+}
+
+func clusterAndMachineTag(clusterName, machineName string) []string {
+	return []string{
+		hvclient.TagKeyClusterName + "=" + clusterName,
+		hvclient.TagKeyMachineName + "=" + machineName,
+	}
+}
+
+// chooseDevice search for an unused device, and then associates the device.
+func (s *Service) chooseDevice(ctx context.Context) (*hv.BareMetalDevice, error) {
+	devices, err := s.scope.HVClient.ListDevices(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("[chooseDevice] ListDevices() failed. machine %q: %w",
+			s.scope.Name(), err)
+	}
+	return chooseAvailableFromList(devices, s.scope.HivelocityMachine.Spec.Type, s.scope.Cluster.Name)
+}
+
+func chooseAvailableFromList(devices []*hv.BareMetalDevice, deviceType infrav1.HivelocityDeviceType, clusterName string) (*hv.BareMetalDevice, error) {
+	for _, device := range devices {
+		// Ignore if associated already
+		if isAssociated(device) {
+			continue
+		}
+
+		// Ignore if has wrong device type
+		dt, err := getDeviceType(device)
+		if err != nil {
+			if errors.Is(err, ErrNoMatchingTagFound) {
+				continue
+			}
+			return nil, fmt.Errorf("[chooseAvailableFromList] getDeviceType() failed: %w", err)
+		}
+
+		if dt != string(deviceType) {
+			continue
+		}
+
+		// Ignore if associated to other cluster
+		cn, found := findAssociatedCluster(device)
+		if found && clusterName != cn {
+			continue
+		}
+		return device, nil
+	}
+	return nil, errNoDeviceAvailable
+}
+
+var errNoDeviceAvailable = fmt.Errorf("no available device found")
+
+// getDeviceType returns the device-type of this BareMetalDevice.
+func getDeviceType(device *hv.BareMetalDevice) (string, error) {
+	deviceType, err := findValueForKeyInTags(device.Tags, hvclient.TagKeyDeviceType)
+	if err != nil {
+		return "", fmt.Errorf("[getDeviceType] DeviceGetTagValue() failed: %w", err)
+	}
+	return deviceType, nil
+}
+
+// isAssociated returns the device-type of this BareMetalDevice.
+func isAssociated(device *hv.BareMetalDevice) bool {
+	deviceType, _ := findValueForKeyInTags(device.Tags, hvclient.TagKeyMachineName)
+	return deviceType != ""
+}
+
+// findAssociatedCluster tries to find a cluster name in the tags. If it finds on, it returns true.
+func findAssociatedCluster(device *hv.BareMetalDevice) (string, bool) {
+	cluster, _ := findValueForKeyInTags(device.Tags, hvclient.TagKeyClusterName)
+	return cluster, cluster != ""
+}
+
+// ErrTooManyTagsFound gets returned, if there are multiple tags with the same key,
+// and the key should be unique.
+var ErrTooManyTagsFound = fmt.Errorf("too many tags found")
```

Comment:
you don't need to export these errors I guess.

Reply:
Additionally, you should put all errors at the top in one var ( ) statement

Reply:
done.